### PR TITLE
[SPARK-44364] [PYTHON] Add support for List[Row] data type for expected

### PR DESCRIPTION
**Pull request by asl3**
### What changes were proposed in this pull request?
This PR adds supported for List[Row] type for the `expected` argument in `assertDataFrameEqual`.

### Why are the changes needed?
The change improves flexibility of the `assertDataFrameEqual` function by allowing for comparison between dataframe and List[Row] types.

### Does this PR introduce _any_ user-facing change?
Yes, the PR modifies the user-facing API `assertDataFrameEqual`.

### How was this patch tested?
Added tests to `python/pyspark/sql/tests/test_utils.py` and `python/pyspark/sql/tests/connect/test_utils.py`

### CODE REVIEW

**Review by ueshin — COMMENTED** [review]
Could you update the expected type hint? Also need to update the doc.

File: python/pyspark/testing/utils.py [hunk 1]
```diff
@@ -382,18 +388,23 @@ def assert_rows_equal(rows1: List[Row], rows2: List[Row]):
         try:
             # rename duplicate columns for sorting
             renamed_df = df.toDF(*[f"_{i}" for i in range(len(df.columns))])
-            renamed_expected = expected.toDF(*[f"_{i}" for i in range(len(expected.columns))])
-
             df = renamed_df.sort(renamed_df.columns).toDF(*df.columns)
-            expected = renamed_expected.sort(renamed_expected.columns).toDF(*expected.columns)
+            if isinstance(expected, List):
+                expected.sort()
```
Comment (ueshin) [inline python/pyspark/testing/utils.py:393]:
```py
expected = sorted(expected)
```

to avoid updating the input.

Comment (ueshin) [reply python/pyspark/testing/utils.py:393]: Btw, I'm wondering Python's `sort` or `sorted` works as same as `df.sort()`. 
We might want to ask users to make it already sorted when they specify a list here.
cc @allisonwang-db

Comment (allisonwang-db) [reply python/pyspark/testing/utils.py:393]: There is a flag  `check_row_orders`. If this flag is False, we will sort both the input data frame and the expected list of Rows.

Comment (allisonwang-db) [reply python/pyspark/testing/utils.py:393]: We can convert the input data frame to a list and sort both lists of rows.

File: python/pyspark/sql/tests/test_utils.py [hunk 2]
```diff
@@ -729,46 +729,6 @@ def test_diff_schema_lens(self):
             message_parameters={"df_schema": df1.schema, "expected_schema": df2.schema},
         )
 
-    def test_assert_equal_maptype(self):
```
Comment (ueshin) [inline python/pyspark/sql/tests/test_utils.py:732]: Shall we keep and update this test to see whether the change on `sort` works for map type or not? If not, we should still show some error.

File: python/pyspark/testing/utils.py [hunk 3]
```diff
@@ -291,7 +293,11 @@ def assertDataFrameEqual(df: DataFrame, expected: DataFrame, check_row_order: bo
                 error_class="UNSUPPORTED_DATA_TYPE",
                 message_parameters={"data_type": type(df)},
             )
-        elif not isinstance(expected, DataFrame) and not isinstance(expected, ConnectDataFrame):
+        elif (
+            not isinstance(expected, DataFrame)
+            and not isinstance(expected, ConnectDataFrame)
+            and not isinstance(expected, List)
```
Comment (HyukjinKwon) [inline python/pyspark/testing/utils.py:299]:
```suggestion
            and not isinstance(expected, list)
```

List is for type hint.

File: python/pyspark/testing/utils.py [hunk 4]
```diff
@@ -235,7 +237,7 @@ def assertDataFrameEqual(df: DataFrame, expected: DataFrame, check_row_order: bo
     df : DataFrame
         The DataFrame that is being compared or tested.
 
-    expected : DataFrame
+    expected : DataFrame or List[Row]
```
Comment (HyukjinKwon) [inline python/pyspark/testing/utils.py:240]:
```suggestion
    expected : DataFrame or list of rows
```

File: python/pyspark/testing/utils.py [hunk 5]
```diff
@@ -221,10 +221,12 @@ def check_error(
         )
 
 
-def assertDataFrameEqual(df: DataFrame, expected: DataFrame, check_row_order: bool = False):
+def assertDataFrameEqual(
+    df: DataFrame, expected: Union[DataFrame, List[Row]], check_row_order: bool = False
```
Comment (HyukjinKwon) [inline python/pyspark/testing/utils.py:225]: `check_row_order` -> `checkRowOrder` or `checkOrder`

**Review by ueshin — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Review by HyukjinKwon — APPROVED** [review]

**Pull request closed without merging**